[5.2] Replace attributes for array keys in validation

themsaid opened the pull request:
As reported in this issue https://github.com/laravel/framework/issues/12548

On array validation setting a custom attribute for `users.*.name` wasn't working, this PR will allow setting an attribute name like this:

```
'attributes' => [
    'users.*.name' => ' User name'
],
```

To get a message like that:

```
The User name must be a string.
```

themsaid committed:
Replace attributes for arrays in validation

taylorotwell commented:
That getCustomAttribute will be quite slow if we have to run Str::is over every custom attribute?

themsaid committed:
some tweaks for performance

themsaid commented:
You are right. I've changed the implementation to use the `$this->implicitAttributes` we added earlier to get the raw attribute, and check for a custom attribute using it.

```
if (isset($this->customAttributes[$rawAttribute])) {
    return $this->customAttributes[$rawAttribute];
}
```

themsaid committed:
if no replacement foudn for the custom attribute by raw name display the attribute

taylorotwell commented:
Merged, but I wonder if it would still be faster to translate all digits to `*` and then check if `array_key_exists` in `implicitAttributes`.

themsaid commented:
The thing here is that array keys might not be digits:

```
$data = [
    'characters' => [
        'humans' => [['name' => 'jon']],
        'white_walkers' => [['name' => null]]
    ]
];

$rules = ['characters.*.*.name' => 'present'];

Validator::make($data, $rules)
```

The attribute name here would be `characters.humans.0.name`.
